Documentation/ecasound.1
.TH "ecasound" "1" "19\&.09\&.2008" "" "Multimedia software"

.PP 
.SH "NAME"
ecasound \- sample editor, multitrack recorder, fx-processor, etc\&.
.PP 
.SH "SYNOPSIS"
\fBecasound\fP [ general_options ] { [ chain_setup ] [ effect_setup ] [ input_setup ] [ output_setup ] }
.PP 
.SH "DESCRIPTION"

.PP 
Ecasound is a software package designed for multitrack audio
processing\&. It can be used for simple tasks like audio playback, 
recording and format conversions, as well as for multitrack effect 
processing, mixing, recording and signal recycling\&. Ecasound supports 
a wide range of audio inputs, outputs and effect algorithms\&. 
Effects and audio objects can be combined in various ways, and their
parameters can be controlled by operator objects like oscillators 
and MIDI-CCs\&. A versatile console mode user-interface is included 
in the package\&.
.PP 
.SH "OPTIONS"

.PP 
Note! All options except those mentioned in \fIecasound options\fP and 
\fIGlobal options\fP, can be used in ecasound chainsetup files (\&.ecs)\&.
.PP 
\fBECASOUND OPTIONS\fP
.PP 
These options are parsed and handled by the ecasound frontbed binary and 
are not passed to backend library\&. This means that these options may
not work in other applications that use ecasound libraries for their
functionality\&.
.PP 
.IP "-c"
Starts ecasound in interactive mode\&. In interactive mode you can
control ecasound with simple commands ("start", "stop", "pause",
etc\&.)\&. See ecasound-iam \&.
.IP 
.IP "-C"
Disables ecasound\&'s interactive mode (see \&'-c\&' and \&'-K\&')\&.

.PP 
.IP "-D"
Print all debug information to stderr (unbuffered, plain output
without ncurses)\&.
.PP 
.IP "-s[:]chainsetup-file"
Create a new chainsetup from file \&'chainsetup-file\&' and add
it to the current session\&. Chainsetup files commonly have 
a filename ending to the \&'\&.ecs\&' extension\&. A chainsetup can 
contain inputs, outputs, chains, effects, controllers -- i\&.e\&.
objects one one specific configuration of audio processing
elements\&. A session, on the other hand, is a collection of 
one or more chainsetups\&. Only one of the chainsetups may be
connected (i\&.e\&. it can be run/processed)\&. But it is possible 
to have another chainsetup select (i\&.e\&. can be configured)
while other one is current connteced (i\&.e\&. running)\&. 
.PP 
.IP "-E "cmd1 [[args] ; cmd2 args ; \&.\&.\&. ; cmdN]""
Execute a set of Ecasound Interactive mode (EIAM) commands
at launch\&. These commands are executed immediately after 
ecasound is started\&. If the command line contains sufficient
options to create a valid chainsetup that will be executed,
the launch commands are executed after the other command
line options are parsed, but before the processing engine
is started\&. Note that this command is a feature of 
the ecasound frontend binary and not supported by 
the library backend\&. This means that other clients may
not support the \&'-E\&' option, and also that the launch
commands are not saved as part of chainsetup or session
state\&.
.PP 
.IP "--daemon"
Enable ecasound\&'s daemon mode\&. If this option is given,
ecasound will allow clients to connect to the running 
ecasound session using a TCP/IP network connection\&.
The clients can both observe and control the session\&. 
.PP 
\fIWarning!\fP As there is no access control implemented,
be sure to block ecasound\&'s port in your firewall 
if the machine running ecasound is connected to 
a public network! Otherwise anyone can connect to 
your ecasound sessions\&.
.PP 
.IP "--daemon-port"
Set the TCP port used by the daemon mode\&. By default
ecasound will use port number \fI2868\fP\&.
.PP 
.IP "--nodaemon"
Disable ecasound\&'s daemon mode\&. This is the default\&.
.PP 
.IP "--keep-running,-K"
Do not exit when processing is finished/stopped\&. Only affects
non-interactive operating mode (see -c/-C)\&.
Option added to ecasound 2\&.4\&.2\&.
.PP 
.IP "--help,-h"
Show this help\&.
.PP 
.IP "--version"
Print version info\&.
.PP 
\fBGLOBAL OPTIONS\fP
.PP 
.IP "-d, -dd, -ddd"
Increase the amount of printed debug messages\&. \fI-d\fP adds 
some verbosity, while \fI-ddd\fP results in very detailed 
output\&.
.IP 
.IP "-d:debug_level"
Set the debug level mask to \&'debug_level\&'\&. This a bitmasked value with
the following classes: errors (1), info (2), subsystems (4), module_names (8),
user_objects (16), system_objects 32, functions (64), continuous (128) and
eiam_return_values (256)\&. Default is 271 (1+2+4+8+256)\&. See sourcode 
documentation for the ECA_LOGGER class for more detailed information\&.
.IP 
.IP "-R[:]path-to-file"
Use ecasound resource file (see ecasoundrc man page) \&'path-to-file\&' as 
the only source of setting resource value\&. Specifying this option
will disable the normal policy of querying both global and user (if exists)
resource files\&.
.IP 
.IP "-q"
Quiet mode, no output\&. Same as \fI-d:0\fP\&.
.IP 
\fBGENERAL CHAINSETUP OPTIONS\fP

.IP 
.IP "-a:chainname1, chainname2, \&.\&.\&."
Selects active signal chains\&. All inputs and outputs following
this \&'-a\&' option are assigned to selected chains (until a new -a
option is specified)\&. When adding effects, controllers and other
chain operators, only one chain can be selected at a time\&. If no -a option 
has been given, chain \&'default\&' is used instead when adding objects\&. 
Chain name \&'all\&' is also reserved\&. It will cause all existing chains
to be selected\&. By giving multiple -a options, you can control to which 
chains effects, inputs and outputs are assigned to\&. Look at the \fBEXAMPLES\fP
section for more detailed info about the usage of this option\&.
.IP 
.IP "-n:name"
Sets the name of chainsetup to \&'name\&'\&. If not specified, defaults
either to "command-line-setup" or to the file name from which
chainsetup was loaded\&. Whitespaces are not allowed\&.
.IP 
.IP "-x"
Truncate outputs\&. All output object are opened in overwrite mode\&. 
Any existing files will be truncated\&.
.IP 
.IP "-X"
Open outputs for updating\&. Ecasound opens all outputs - if target
format allows it - in readwrite mode\&.
.IP 
.IP "-z:feature"
Enables \&'feature\&'\&. Most features can be disabled using notation 
\fI-z:nofeature\fP\&. \&'-z:db,dbsize\&' enables double-buffering for audio 
objects that support it (dbsize=0 for default, otherwise buffer
size in sample frames)\&. \&'-z:nodb\&' disables double-buffering\&. 
\&'-z:intbuf\&' and \&'-z:nointbuf\&' control whether extra internal buffering 
is allowed for realtime devices\&. Disabling this can reduce 
latency times in some situations\&. With \&'-z:xruns\&', processing will be 
halted if an under/overrun occurs\&. \&'-z:multitrack\&' and
\&'z:nomultitrack\&' can be used to force ecasound to enable or disable
multitrack-mode\&. In rare cases you may want to explicitly specify 
the recording offset with \&'-z:multitrack,offset-in-samples\&'\&. The
offset is the amount of samples skipped when recording from 
real-time inputs\&. \&'-z:psr\&' enables the \fIprecise-sample-rates\fP mode 
for OSS-devices\&. \&'-z:mixmode,sum\&' enables mixing mode where channels
are mixed by summing all channels\&. The default is \&'-z:mixmode,avg\&',
in which channels are mixed by averaging\&. Mixmode selection was first
added to ecasound 2\&.4\&.0\&.
See ecasoundrc man page\&.
.IP 
\fBCHAINSETUP BUFFERING AND PERFORMANCE OPTIONS\fP

.IP 
.IP "-B:buffering_mode"
Selects the default buffering mode\&. Mode is one of: \&'auto\&' (default),
\&'nonrt\&', \&'rt\&', \&'rtlowlatency\&'\&.
.IP 
.IP "-b:buffer size"
Sets the size of buffer in samples (must be an exponent of 2)\&. This
is quite an important option\&. For real-time processing, you should
set this as low as possible to reduce the processing delay\&. Some
machines can handle buffer values as low as 64 and 128\&. In some
circumstances (for instance when using oscillator envelopes) small
buffer sizes will make envelopes act more smoothly\&. When not processing
in real-time (all inputs and outputs are normal files), values between
512 - 4096 often give better results\&. Default is 1024\&.
.IP 
.IP "-r:sched_priority"
Use realtime scheduling policy (SCHED_FIFO)\&. This is impossible if 
ecasound doesn\&'t have root priviledges\&. Beware! This gives better 
performance, but can cause total lock-ups if something goes wrong\&.
The \&'sched_priority\&' can be omitted (0=omitted)\&. If given, 
this is the static priority to the highest priority ecasound thread\&.
Other ecasound threads run with priority \&'sched_priority-1\&.\&.\&.n\&'\&.
Value \&'-1\&' can be used to disable raised-priority mode\&.
.IP 
.IP "-z:feature"
Relevant features are -z:db,xxx (-z:nodb) and -z:intbuf (-z:nointbuf)\&.
See section \fIGeneral chainsetup options\fP for details\&.
.IP 
\fBPROCESSING CONTROL\fP
.IP "-t:seconds"
Sets processing time in seconds (doesn\&'t have to be an integer value)\&.
If processing time isn\&'t set, engine stops when all inputs are 
finished\&. This option is equivalent to the \&'cs-set-length\&' EIAM
command\&. A special-case value of \&'-1\&' will set the chainsetup length 
according to the longest input object\&.
.IP 
.IP "-tl"
Enables looping\&. When processing is finished, engine will start 
again from beginning\&. This option is equivalent to the \&'cs-loop\&' 
EIAM command\&.
.IP 
\fBINPUT/OUTPUT SETUP\fP
.PP 
See ecasound user\&'s guide for 
more detailed documentation\&.
.PP 
.IP "-G:mgrtype,optstring"
Sets options for audio object manager type \&'mgrtype\&'\&. 
For available options, see "OBJECT TYPE SPECIFIC NOTES" below\&.
.IP 
.IP "-f:sample_format,channel,sample-rate,interleaving"
Sets default sampling parameters\&. These are used for all following
input and output files or until another -f is specified\&. If no -f
option is present, ecasound will use the default audio parameters (see
ecasoundrc man page)\&. Notice that when opening existing audio objects, 
either for input or output, the default audio parameters set with -f are
ignored if objects provide sufficient header information (as
is the case for wav, aiff, etc formats)\&. For output objects, the
-x option can be used to completely overwrite existing files 
(in other words, with -x, default audio parameters set with -f are used)\&.
.IP 
Sample format is given as a a formatted string\&. The first letter is 
either "u", "s" and "f" (unsigned, signed, floating point)\&. The 
following number specifies sample size in bits\&. If sample is 
little endian, "_le" is added to the end\&. Similarly if big endian, 
"_be" is added\&. If endianess is not specified, host byte-order is used\&. 
Currently supported formats are "u8" (same as "8"), "s16_le" (same 
as "16"), "s16_be", "s24_le", "s24_be", "s32_le", "s32_be", "f32_le" 
and "f32_be"\&.
.IP 
The 4th parameter \&'interleaving\&' should either be \&'i\&' (default) for
interleaved stream format, or \&'n\&' for noninterleaved\&.
.IP 
.IP "-y:seconds"
Sets starting position for last specified input/output\&. If 
you need more flexible control over audio objects, you should
use the \fI\&.ewf\fP format\&.
.IP 
.IP "-i[:]input-file-or-device[,params]"
Specifies a new input source that is connected to all selected chains\&.
Connecting multiple inputs to the same chain isn\&'t possible\&. Input
can be a a file, device or some other audio object (see below)\&. 
If the input is a file, its type is determined using the file name 
extension\&. If the object name contains any commas, the name must be 
enclosed in backquotes to avoid confusing the parser\&. Currently supported 
formats are RIFF WAVE files (\&.wav), audio-cd tracks (\&.cdr), ecasound 
EWF files (\&.ewf), RAW audio data (\&.raw) and MPEG files (\&.mp2,\&.mp3)\&. Also, 
formats supported by the SGI audiofile library: AIFF (\&.aiff, \&.aifc, \&.aif) and 
Sun/NeXT audio files (\&.au, \&.snd)\&. MikMod is also supported (\&.xm, \&.mod, \&.s3m, 
\&.it, etc)\&. MIDI files (\&.mid) are supported using Timidity++\&. Similarly
Ogg Vorbis (\&.ogg) can be read and written if ogg123 and vorbize tools
are installed, FLAC files (\&.flac) with flac command-line tools or using
libsndfile, and AAC files (\&.aac/\&.m4a/\&.mp4) with faad2/faac tools\&. Supported 
realtime devices are OSS audio devices (/dev/dsp*), ALSA audio and loopback 
devices and JACK audio subsystem\&. If no inputs are specified, the first 
non-option (doesn\&'t start with \&'-\&') command line argument is considered 
to be an input\&.
.IP 
.IP "-o[:]output-file-or-device[,params]"
Works in the same way as the -i option\&. If no outputs are specified,
the default output device is used (see ~/\&.ecasoundrc)\&. If the object 
name contains any commas, the name must be enclosed in backquotes to 
avoid confusing the parser\&. Note, you can\&'t output to module formats 
supported by MikMod (this should be obvious)\&.
.IP 
\fIOBJECT TYPE SPECIFIC NOTES\fP
.IP "ALSA devices"
When using ALSA drivers, instead of a device filename, you need to
use the following option syntax: \fB-i[:]alsa,pcm_device_name\fP\&.
.IP 
.IP "ALSA direct-hw and plugin access"
It\&'s also possible to use a specific card and device combination
using the following notation: \fB-i[:]alsahw,card_number,device_number,subdevice_number\fP\&.
Another option is the ALSA PCM plugin layer\&. It works just like 
the normal ALSA pcm-devices, but with automatic channel count and 
sample format conversions\&. Option syntax is 
\fB-i[:]alsaplugin,card_number,device_number,subdevice_number\fP\&.
.IP 
.IP "aRts input/output"
If enabled at compile-time, ecasound supports audio input and 
output using aRts audio server\&. Option syntax is \fB-i:arts\fP,
\fB-o:arts\fP\&.
.IP 
.IP "Audio file sequecing"
Ecasound provides a set of special audio object types that 

ecasound/ecasound.cpp
// ------------------------------------------------------------------------
// ecasound.cpp: Console mode user interface to ecasound.
// Copyright (C) 2002-2008 Kai Vehmanen
//
// Attributes:
//     eca-style-version: 3 (see Ecasound Programmer's Guide)
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307  USA
// ------------------------------------------------------------------------

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <cstdlib>
#include <iostream>
#include <string>
#include <vector>

#include <signal.h>    /* POSIX: various signal functions */
#include <unistd.h>    /* POSIX: sleep() */

#include <kvu_dbc.h>
#include <kvu_com_line.h>
#include <kvu_utils.h>

#include <eca-control.h>
#include <eca-error.h>
#include <eca-logger.h>
#include <eca-session.h>
#include <eca-version.h>

#include "eca-comhelp.h"
#include "eca-console.h"
#include "eca-curses.h"
#include "eca-neteci-server.h"
#include "eca-plaintext.h"
#include "textdebug.h"
#include "ecasound.h"

/**
 * Static/private function definitions
 */

static void ecasound_create_eca_objects(struct ecasound_state* state, COMMAND_LINE& cline);
static void ecasound_launch_daemon(struct ecasound_state* state);
static int ecasound_pass_at_launch_commands(struct ecasound_state* state);
static void ecasound_main_loop(struct ecasound_state* state);
void ecasound_parse_command_line(struct ecasound_state* state, 
				 const COMMAND_LINE& clinein,
				 COMMAND_LINE* clineout);
static void ecasound_print_usage(void);
static void ecasound_print_version_banner(void);
static void ecasound_setup_signals(struct ecasound_state* state);

extern "C" {
  static void ecasound_exit_cleanup(void);
  static void* ecasound_signal_watchdog_thread(void* arg); 
  static void ecasound_signal_handler(int signal);
}

static struct ecasound_state ecasound_state_global = 
  { 0,        /* ECA_CONSOLE* */
    0,        /* ECA_CONTROL */
    0,        /* ECA_LOGGER_INTERFACE */
    0,        /* ECA_NETECI_SERVER */
    0,        /* ECA_SESSION */
    0,        /* launchcmds, std::vector<std::string> */
    0,        /* pthread_t - daemon_thread */
    0,        /* pthread_mutex_t - lock */
    0,        /* sig_wait_t - exit_request */
    0,        /* sigset_t */
    ECASOUND_RETVAL_SUCCESS, /* int - return value */
    2868,     /* int - default daemon mode TCP-port */
    false,    /* daemon mode */
    false,    /* keep_running mode */
    false,    /* cerr-output-only mode */
    false,    /* interactive mode */
    false     /* quiet mode */
  };

static sig_atomic_t ecasound_cleanup_done = 0;
static sig_atomic_t ecasound_normal_exit = 0;
static sig_atomic_t ecasound_watchdog_active = 0;

/**
 * Namespace imports
 */
using namespace std;

/**
 * Function definitions
 */
int main(int argc, char *argv[])
{
  struct ecasound_state* state = &ecasound_state_global;

  /* 1. setup signals and the signal watchdog thread */
  ecasound_setup_signals(state);

  /* 2. parse command-line args */
  COMMAND_LINE* cline = new COMMAND_LINE(argc, argv);
  COMMAND_LINE* clineout = new COMMAND_LINE();
  ecasound_parse_command_line(state, *cline, clineout); 
  delete cline; cline = 0;

  /* 3. create console interface */
  if (state->retval == ECASOUND_RETVAL_SUCCESS) {

#if defined(ECA_PLATFORM_CURSES) 
    if (state->quiet_mode != true &&
	state->cerr_output_only_mode != true) {
      state->console = new ECA_CURSES();
      state->logger = new TEXTDEBUG();
      ECA_LOGGER::attach_logger(state->logger);
    }
    else 
#endif
      {
	ostream* ostr = (state->cerr_output_only_mode == true) ? &cerr : &cout;
	state->console = new ECA_PLAIN_TEXT(ostr);
      }
    
    if (state->quiet_mode != true) {
      /* 4. print banner */
      state->console->print_banner();
    }

    /* 5. set default debug levels */
    ECA_LOGGER::instance().set_log_level(ECA_LOGGER::errors, true);
    ECA_LOGGER::instance().set_log_level(ECA_LOGGER::info, true);
    ECA_LOGGER::instance().set_log_level(ECA_LOGGER::subsystems, true);
    ECA_LOGGER::instance().set_log_level(ECA_LOGGER::eiam_return_values, true);
    ECA_LOGGER::instance().set_log_level(ECA_LOGGER::module_names, true);
    
    /* 6. create eca objects */
    ecasound_create_eca_objects(state, *clineout);
    delete clineout; clineout = 0;

    /* 7. start ecasound daemon */
    if (state->retval == ECASOUND_RETVAL_SUCCESS) {
      if (state->daemon_mode == true) {
	ecasound_launch_daemon(state);
      }
    }

    /* 8. pass launch commands */
    ecasound_pass_at_launch_commands(state);

    /* 9. start processing */
    if (state->retval == ECASOUND_RETVAL_SUCCESS) {
      ecasound_main_loop(state);
    }
  }

  if (state->daemon_mode == true) {
    /* wait until daemon thread has exited */
    if (state->interactive_mode == true) {
      state->exit_request = 1;
    }
    pthread_join(*state->daemon_thread, NULL);
  }

  /* note: if we exist due to a signal, we never reach 
   *       the end of main() */

  /* note: we prefer to run the cleanup routines before 
   *       returning from main; there have been problems with
   *       dynamic libraries and libecasound cleanup routines
   *       when run from an atexit() handler
   */
  ecasound_normal_exit = 1;
  ecasound_exit_cleanup();

  // cerr << endl << "ecasound: main() exiting..." << endl << endl;

  DBC_CHECK(state->retval == ECASOUND_RETVAL_SUCCESS ||
	    state->retval == ECASOUND_RETVAL_INIT_FAILURE ||
	    state->retval == ECASOUND_RETVAL_START_ERROR ||
	    state->retval == ECASOUND_RETVAL_RUNTIME_ERROR);

  return state->retval;
}

/**
 * Cleanup routine that is run after either exit()
 * is called or ecasound returns from its main().
 */
void ecasound_exit_cleanup(void)
{
  struct ecasound_state* state = &ecasound_state_global;

  // cerr << endl << "ecasound: atexit cleanup" << endl;

  if (ecasound_cleanup_done == 0) {
    ecasound_cleanup_done  = 1;  
    if (state->control != 0) {
      if (state->control->is_running() == true) {
	state->control->stop_on_condition();
      }
      if (state->control->is_connected() == true) {
	state->control->disconnect_chainsetup();
      }
    }

    DBC_CHECK(ecasound_normal_exit == 1);

    if (state->control != 0) { delete state->control; state->control = 0; }
    if (state->session != 0) { delete state->session; state->session = 0; }

    if (state->launchcmds != 0) { delete state->launchcmds; state->launchcmds = 0; }
    if (state->eciserver != 0) { delete state->eciserver; state->eciserver = 0; }
    if (state->console != 0) { delete state->console; state->console = 0; }
    if (state->daemon_thread != 0) { delete state->daemon_thread; state->daemon_thread = 0; }
    if (state->lock != 0) { delete state->lock; state->lock = 0; }
    if (state->signalset != 0) { delete state->signalset; state->signalset = 0; }
  }

  // cerr << "ecasound: exit cleanup done." << endl << endl;
}

/**
 * Enters the main processing loop.
 */
void ecasound_create_eca_objects(struct ecasound_state* state, 
				 COMMAND_LINE& cline)
{
  DBC_REQUIRE(state != 0);
  DBC_REQUIRE(state->console != 0);

  try {
    state->session = new ECA_SESSION(cline);
    state->control = new ECA_CONTROL(state->session);

    DBC_ENSURE(state->session != 0);
    DBC_ENSURE(state->control != 0);
  }
  catch(ECA_ERROR& e) {
    state->console->print("---\necasound: ERROR: [" + e.error_section() + "] : \"" + e.error_message() + "\"\n");
    state->retval = ECASOUND_RETVAL_INIT_FAILURE;
  }
}

/**
 * Launches a background daemon that allows NetECI 
 * clients to connect to the current ecasound
 * session.
 */
void ecasound_launch_daemon(struct ecasound_state* state)
{
  DBC_REQUIRE(state != 0);
  // DBC_REQUIRE(state->console != 0);

  // state->console->print("ecasound: starting the NetECI server.");

  state->daemon_thread = new pthread_t;
  state->lock = new pthread_mutex_t;
  pthread_mutex_init(state->lock, NULL);
  state->eciserver = new ECA_NETECI_SERVER(state);

  int res = pthread_create(state->daemon_thread, 
			   NULL,
			   ECA_NETECI_SERVER::launch_server_thread, 
			   reinterpret_cast<void*>(state->eciserver));
  if (res != 0) {
    cerr << "ecasound: Warning! Unable to create daemon thread." << endl;
    delete state->daemon_thread;  state->daemon_thread = 0;
    delete state->lock;  state->lock = 0;
    delete state->eciserver; state->eciserver = 0;
  }

  // state->console->print("ecasound: NetECI server started");
}

static int ecasound_pass_at_launch_commands(struct ecasound_state* state)
{
  if (state->launchcmds) {
    std::vector<std::string>::const_iterator p = state->launchcmds->begin();

    while(p != state->launchcmds->end()) {
      state->control->command(*p);
      state->control->print_last_value();
      ++p;
    }
  }

  return 0;
}

/**
 * The main processing loop.
 */
void ecasound_main_loop(struct ecasound_state* state)
{
  DBC_REQUIRE(state != 0);
  DBC_REQUIRE(state->console != 0);

  ECA_CONTROL* ctrl = state->control;

  if (state->interactive_mode == true) {

    while(state->exit_request == 0) {
      state->console->read_command("ecasound ('h' for help)> ");
      const string& cmd = state->console->last_command();
      if (cmd.size() > 0 && state->exit_request == 0) {

	if (state->daemon_mode == true) {
	  int res = pthread_mutex_lock(state->lock);
	  DBC_CHECK(res == 0);
	}

	ctrl->command(cmd);
	ctrl->print_last_value();

	if (state->daemon_mode == true) {
	  int res = pthread_mutex_unlock(state->lock);
	  DBC_CHECK(res == 0);
	}

	if (cmd == "quit" || cmd == "q") {
	  state->console->print("---\necasound: Exiting...");
	  state->exit_request = 1;
	  ECA_LOGGER::instance().flush();
	}
      }
    }
  }
  else {
    /* non-interactive mode */

    if (state->daemon_mode == true) {
      int res = pthread_mutex_lock(state->lock);
      DBC_CHECK(res == 0);
    }

    if (ctrl->is_selected() == true && 
	ctrl->is_valid() == true) {
      ctrl->connect_chainsetup();
    }


    if (ctrl->is_connected() == true) {
      if (!state->exit_request) {
	int res = ctrl->run(!state->keep_running_mode);
	if (res < 0) {
	  state->retval = ECASOUND_RETVAL_RUNTIME_ERROR;
	  cerr << "ecasound: Warning! Errors detected during processing." << endl;
	}
      }
    }
    else {
      ctrl->print_last_value();
      state->retval = ECASOUND_RETVAL_START_ERROR;
    }

    if (state->daemon_mode == true) {
      int res = pthread_mutex_unlock(state->lock);
      DBC_CHECK(res == 0);
    }
  }
  // cerr << endl << "ecasound: mainloop exiting..." << endl;
}

/**
 * Parses the command lines options in 'cline'.
 */
void ecasound_parse_command_line(struct ecasound_state* state, 
				 const COMMAND_LINE& cline,
				 COMMAND_LINE* clineout)
{
  if (cline.size() < 2) {
    ecasound_print_usage();
    state->retval = ECASOUND_RETVAL_INIT_FAILURE;
  }
  else {
   cline.begin();
    while(cline.end() != true) {
      if (cline.current() == "-o:stdout" ||
	  cline.current() == "stdout" ||
	  cline.current() == "-d:0" ||
	  cline.current() == "-q") {

	state->quiet_mode = true;
	/* pass option to libecasound */
	clineout->push_back(cline.current());
      } 

      else if (cline.current() == "-c") {
	state->interactive_mode = true;
      }

      else if (cline.current() == "-C") {
	state->interactive_mode = false;
      }

      else if (cline.current() == "-D") {
	state->cerr_output_only_mode = true;
      }
      
      else if (cline.current() == "--daemon") {
	state->daemon_mode = true;
      }

      else if (cline.current().find("-E") != string::npos) {
	cline.next();
	if (cline.end() != true) {
	  state->launchcmds = 
	    new std::vector<std::string>
 	      (kvu_string_to_vector(cline.current(), ';'));
	}
      }

      else if (cline.current().find("--daemon-port") != string::npos) {
	std::vector<std::string> argpair = 
	  kvu_string_to_vector(cline.current(), '=');
	if (argpair.size() > 1) {
	  /* --daemon-port=XXXX */
	  state->daemon_port = atoi(argpair[1].c_str());
	}
      }


      else if (cline.current() == "--nodaemon") {
	state->daemon_mode = false;
      }

      else if (cline.current() == "-h" ||
	       cline.current() == "--help") {
	ecasound_print_usage();
	state->retval = ECASOUND_RETVAL_INIT_FAILURE;
	break;
      }

      else if (cline.current() == "-K" ||
	       cline.current() == "--keep-running") {
	state->keep_running_mode = true;
      }

      else if (cline.current() == "--version") {
	ecasound_print_version_banner();
	state->retval = ECASOUND_RETVAL_INIT_FAILURE;
	break;
      }
      
      else {
	/* pass rest of the options to libecasound */
	clineout->push_back(cline.current());
      }

      cline.next();
    }
  }
}

void ecasound_print_usage(void)
{
  cout << ecasound_parameter_help();
}

void ecasound_print_version_banner(void)
{
  cout << "ecasound v" << ecasound_library_version << endl;
  cout << "Copyright (C) 1997-2008 Kai Vehmanen and others." << endl;
  cout << "Ecasound comes with ABSOLUTELY NO WARRANTY." << endl;
  cout << "You may redistribute copies of ecasound under the terms of the GNU" << endl;
  cout << "General Public License. For more information about these matters, see" << endl; 
  cout << "the file named COPYING." << endl;
}

static void ecasound_signal_handler(int signal)
{
#if defined(HAVE_SIGPROCMASK) || defined(HAVE_PTHREAD_SIGMASK)
  if (ecasound_watchdog_active &&
      ecasound_state_global.exit_request == 0) {
    cerr << "(ecasound-watchdog) WARNING: ecasound_signal_handler entered, this should _NOT_ happen!";
    cerr << " pid=" << getpid() << endl;
  }
#endif

  if (ecasound_watchdog_active &&
      ecasound_state_global.exit_request) {
    cerr << "(ecasound-watchdog) WARNING: Signal received during cleanup, exiting immediately.\n";
    exit(ECASOUND_RETVAL_RUNTIME_ERROR);
  }
}

/**
 * Sets up a signal mask with sigaction() that blocks 
 * all common signals, and then launces a signal watchdog
 * thread that waits on the blocked signals using
 * sigwait(). 
 * 
 * This design causes all non-fatal termination signals
 * to be routed through a single thread. This signal watchdog
 * in turn performs a clean exit upon receiving a signal.
 * Without this setup, interactions between threads when handling 
 * would be harder to control (especially considering that
 * ecasound needs to work on various different platforms).
 */
void ecasound_setup_signals(struct ecasound_state* state)
{
  pthread_t watchdog;
  sigset_t* signalset;

  /* man pthread_sigmask:
   *  "...signal actions and signal handlers, as set with
   *   sigaction(2), are shared between all threads"
   */

  /* handle the following signals explicitly */
  signalset = new sigset_t;
  state->signalset = signalset;
  sigemptyset(signalset);
  sigaddset(signalset, SIGTERM);
  sigaddset(signalset, SIGINT);
  sigaddset(signalset, SIGHUP);
  sigaddset(signalset, SIGPIPE);
  sigaddset(signalset, SIGQUIT);

  /* create a dummy signal handler */
  struct sigaction blockaction;
  blockaction.sa_handler = ecasound_signal_handler;
  sigemptyset(&blockaction.sa_mask);
  blockaction.sa_flags = 0;

  /* attach the dummy handler to the following signals */
  sigaction(SIGTERM, &blockaction, 0);
  sigaction(SIGINT, &blockaction, 0);
  sigaction(SIGHUP, &blockaction, 0);
  sigaction(SIGPIPE, &blockaction, 0);
  sigaction(SIGQUIT, &blockaction, 0);

#ifdef __FreeBSD__
  /* ignore signals instead of passing them to our handler */
  blockaction.sa_handler = SIG_IGN;
  sigaction(SIGFPE, &blockaction, 0);
#endif

  int res = pthread_create(&watchdog, 
			   NULL, 
			   ecasound_signal_watchdog_thread, 
			   reinterpret_cast<void*>(state));
  if (res != 0) {
    cerr << "ecasound: Warning! Unable to create watchdog thread." << endl;
  }

  /* block all signals in 'signalset' (see above) */
#if defined(HAVE_PTHREAD_SIGMASK)
  pthread_sigmask(SIG_BLOCK, signalset, NULL);
#elif defined(HAVE_SIGPROCMASK)
  sigprocmask(SIG_BLOCK, signalset, NULL);
#endif
}

/**
 * Runs a watchdog thread that centrally catches signals that
 * will cause ecasound to exit.
 */
void* ecasound_signal_watchdog_thread(void* arg)
{
  int signalno = 0;
  struct ecasound_state* state = reinterpret_cast<struct ecasound_state*>(arg);

  // cerr << "Watchdog-thread created, pid=" << getpid() << "." << endl;

#ifdef HAVE_SIGWAIT

  /* step: block execution until a signal received 
   **********************************************/

  ecasound_watchdog_active = 1;

# if defined(HAVE_PTHREAD_SIGMASK)
  pthread_sigmask(SIG_BLOCK, state->signalset, NULL);
# elif defined(HAVE_SIGPROCMASK)
  /* the set of signals must be blocked before entering sigwait() */
  sigprocmask(SIG_BLOCK, state->signalset, NULL);
# endif

  sigwait(state->signalset, &signalno);

  cerr << endl << "(ecasound-watchdog) Received signal " << signalno << ". Cleaning up and exiting..." << endl;

  /* step: use pause() instead (alternative to sigwait()) 
  ********************************************************/

#elif HAVE_PAUSE /* !HAVE_SIGWAIT */

  /* note: with pause() we don't set 'ecasound_watchdog_active' as
   * it is normal to get signals when watchdog is running */
  pause();

  cerr << endl << "(ecasound-watchdog) Received signal and returned from pause(). Cleaning up and Exiting..." << Endl;

#else /* !HAVE_PAUSE */

  /* 2. If proper signal handling is not possible, stop compilation.
   *****************************************************************/

# error "Neither sigwait() or pause() Is available, unable to continue."

#endif

  /* step: signal the mainloop that process should terminate */
  state->exit_request = 1;

  /* step: unblock signals after process termination has 
   *       been started */

#ifdef HAVE_SIGWAIT
# if defined(HAVE_PTHREAD_SIGMASK)
  pthread_sigmask(SIG_UNBLOCK, state->signalset, NULL);
# elif defined(HAVE_SIGPROCMASK)
  /* the set of signals must be blocked before entering sigwait() */
  sigprocmask(SIG_UNBLOCK, state->signalset, NULL);
# endif
#endif

  /* step: in case mainloop is blocked running a batch job, we signal
   *       the engine thread directly and force it to terminate */
  if (state->interactive_mode != true &&
      state->control)
    state->control->quit_async();

  while(ecasound_normal_exit != 1) {
    /* sleep for one 200ms */
    kvu_sleep(0, 200000000);

    /* note: A race condition exists between ECA_CONTROL_BASE
     *       quit_async() and run(): if quit_async() is called
     *       after run() has been entered, but before run()
     *       has managed to start the engine, it is possible engine
     *       may still be started. 
     * 
     *       Thus we will keep checking the engine status until 
     *       shutdown is really completed. 
     *
     *       For robustness, this check is also done when in
     *       interactive mode (in case the mainloop does not for
     *       some reason react to our exit request).
     */
    if (state->control) {
      if (state->control->is_engine_started() == true) {
	state->control->quit_async();
      }
    }

  }
    
  ecasound_watchdog_active = 0;

  // cerr << endl << "ecasound: watchdog thread exiting..." << endl;

  return 0;
}
